api/datasets.go

```diff
@@ -646,7 +646,13 @@ func (h DatasetHandlers) bodyHandler(w http.ResponseWriter, r *http.Request) {
 	}
 
 	if err := repo.CanonicalizeDatasetRef(h.repo, &d); err != nil {
-		util.WriteErrResponse(w, http.StatusInternalServerError, err)
+		// TODO: look through lib.LookupBody and see if we can refactor so it takes a datasetRef rather than just a path.
```

Comment:
This bug is my fault (sorry!), but this is not the right fix. The semantics of CanonicalizeDatasetRef are weird because it's now basically doing two things. If the err returned is equal to repo.ErrNotFound, it means the ref was canonicalized but the dataset doesn't exist in the local repo. I made this change but forgot to update all the callers (whoops!). The correct thing to do here is to only send StateInternalServerError if err != nil && err != repo.ErrNotFound. As long as the dataset is remote, the rest of this code will handle it just fine by continuing onto LookupBody.